### ISSUE TITLE
Game crash on `updateSkeleton` - out of bounds a bit deeper in

### ISSUE THREAD

**Issue by Cervator** [MovingBlocks/Terasology#4362]
### What you were trying to do

Test MetalRenegades. Flew fast with `ghost` + `hspeed`

Maybe we can harden or better log some of the related code?

### What actually happened

Crashed after less than a minute - something seemingly went wrong when a creature spawned and needed to be updated?

Seemingly hit some ugly chunk errors afterward as well, unless there's a possibility that was the trigger - only one millisecond later than the skeleton update, and likely in different threads?

### How to reproduce

* Start MetalRenegades
* `hspeed` and fly around fast
* Get lucky and crash with this message?

### Log details and game version

`develop` branch current as of when this issue was submitted

```
...
21:13:37.251 [main] INFO  o.t.logic.console.ConsoleImpl - [CONSOLE] Ghost mode toggled
21:13:38.310 [main] INFO  o.t.logic.console.ConsoleImpl - [CONSOLE] High-speed mode activated
21:13:41.176 [main] ERROR o.t.p.t.TypeHandlerLibrary - Unsupported field: 'DynamicCities:TreeFacet.relData'
21:13:41.264 [main] WARN  o.t.l.c.ServerCharacterPredictionSystem - Received too much input from EntityRef{id = 522, netId = 0, prefab = 'WildAnimals:sheep'}, dropping input. Delta difference: 39.0
21:13:41.988 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.ArrayIndexOutOfBoundsException: 2410
        at gnu.trove.impl.hash.TLongHash.indexRehashed(TLongHash.java:233)
        at gnu.trove.impl.hash.TLongHash.index(TLongHash.java:219)
        at gnu.trove.map.hash.TLongObjectHashMap.get(TLongObjectHashMap.java:212)
        at org.terasology.persistence.internal.EntitySetDeltaRecorder.getOrCreateEntityDeltaFor(EntitySetDeltaRecorder.java:83)
        at org.terasology.persistence.internal.EntitySetDeltaRecorder.onEntityComponentChange(EntitySetDeltaRecorder.java:67)
        at org.terasology.persistence.internal.ReadWriteStorageManager.onEntityComponentChange(ReadWriteStorageManager.java:519)
        at org.terasology.entitySystem.entity.internal.PojoEntityManager.notifyComponentChanged(PojoEntityManager.java:738)
        at org.terasology.entitySystem.entity.internal.PojoEntityManager.saveComponent(PojoEntityManager.java:627)
        at org.terasology.entitySystem.entity.internal.BaseEntityRef.saveComponent(BaseEntityRef.java:183)
        at org.terasology.rendering.logic.SkeletonRenderer.updateSkeleton(SkeletonRenderer.java:208)
        at org.terasology.rendering.logic.SkeletonRenderer.updateSkeletalMeshOfEntity(SkeletonRenderer.java:171)
        at org.terasology.rendering.logic.SkeletonRenderer.update(SkeletonRenderer.java:110)
        at org.terasology.engine.modes.StateIngame.update(StateIngame.java:171)
        at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:470)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:433)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:409)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:13:41.988 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
21:13:41.989 [Chunk-Processing-Reactor] ERROR o.t.w.c.p.ChunkProcessingPipeline - Cannot run java.util.concurrent.ExecutorCompletionService$QueueingFuture@1bdab446  because queue is full
21:13:41.992 [Chunk-Processing-Reactor] ERROR o.t.w.c.p.ChunkProcessingPipeline - Reactor thread was interrupted
java.lang.InterruptedException: null
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireInterruptibly(AbstractQueuedSynchronizer.java:1220)
        at java.util.concurrent.locks.ReentrantLock.lockInterruptibly(ReentrantLock.java:335)
        at java.util.concurrent.PriorityBlockingQueue.take(PriorityBlockingQueue.java:545)
        at java.util.concurrent.ExecutorCompletionService.take(ExecutorCompletionService.java:193)
        at org.terasology.world.chunks.pipeline.ChunkProcessingPipeline.chunkTaskHandler(ChunkProcessingPipeline.java:103)
        at java.lang.Thread.run(Thread.java:745)
21:13:42.246 [main] ERROR o.t.l.debug.ChunkEventErrorLogger - Unload event for not loaded chunk (-8, 0, 1)
21:13:42.258 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
21:13:42.280 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
21:13:42.291 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Player65978
21:13:42.299 [main] INFO  o.t.logic.console.ConsoleImpl - [NOTIFICATION] Player "Player65978" has left the game
21:13:42.300 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
21:13:42.300 [main] WARN  o.t.r.world.RenderableWorldImpl - Warning: onChunkLoaded called for a null chunk!
21:13:42.301 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelWorldSystem.onNewChunk(VoxelWorldSystem.java:151)
        at org.terasology.physics.engine.VoxelWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:13:42.301 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelLiquidWorldSystem.onNewChunk(VoxelLiquidWorldSystem.java:152)
        at org.terasology.physics.engine.VoxelLiquidWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:13:42.454 [Engine-Task-Pool] INFO  o.t.rendering.opengl.ScreenGrabber - Screenshot saved to saves\Game 13\previews\1.jpg!
21:13:43.127 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
21:13:43.134 [Thread-4] INFO  o.t.e.s.d.DiscordRPCSubSystem - Discord RPC >> Disconnected!
```

**Comment by Cervator** [comment]
Maybe it is something else ... same exception, different stack, same exact conditions in-game:

```
21:29:58.686 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
21:30:00.526 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.ArrayIndexOutOfBoundsException: 319
        at gnu.trove.impl.hash.TLongHash.indexRehashed(TLongHash.java:233)
        at gnu.trove.impl.hash.TLongHash.index(TLongHash.java:219)
        at gnu.trove.map.hash.TLongObjectHashMap.get(TLongObjectHashMap.java:212)
        at org.terasology.entitySystem.entity.internal.ComponentTable.get(ComponentTable.java:41)
        at org.terasology.entitySystem.entity.internal.PojoEntityManager.lambda$getComponent$7(PojoEntityManager.java:538)
        at java.util.Optional.map(Optional.java:215)
        at org.terasology.entitySystem.entity.internal.PojoEntityManager.getComponent(PojoEntityManager.java:538)
        at org.terasology.entitySystem.entity.internal.BaseEntityRef.getComponent(BaseEntityRef.java:160)
        at org.terasology.logic.location.LocationComponent.getWorldPosition(LocationComponent.java:213)
        at org.terasology.logic.location.LocationComponent.getWorldPosition(LocationComponent.java:201)
        at org.terasology.logic.location.LocationComponent.getWorldPosition(LocationComponent.java:190)
        at org.terasology.logic.players.LocalPlayer.getViewPosition(LocalPlayer.java:201)
        at org.terasology.logic.players.LocalPlayer.getViewPosition(LocalPlayer.java:185)
        at org.terasology.rendering.cameras.ViewFrustum.intersects(ViewFrustum.java:161)
        at org.terasology.rendering.cameras.ViewFrustum.intersects(ViewFrustum.java:153)
        at org.terasology.rendering.cameras.Camera.hasInSight(Camera.java:259)
        at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:396)
        at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:392)
        at org.terasology.rendering.world.RenderableWorldImpl.queueVisibleChunks(RenderableWorldImpl.java:320)
        at org.terasology.rendering.world.WorldRendererImpl.preRenderUpdate(WorldRendererImpl.java:320)
        at org.terasology.rendering.world.WorldRendererImpl.render(WorldRendererImpl.java:350)
        at org.terasology.engine.modes.StateIngame.render(StateIngame.java:206)
        at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:196)
        at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:479)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:433)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:409)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:30:00.526 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
21:30:00.527 [Chunk-Processing-Reactor] ERROR o.t.w.c.p.ChunkProcessingPipeline - Reactor thread was interrupted
java.lang.InterruptedException: null
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2048)
        at java.util.concurrent.PriorityBlockingQueue.take(PriorityBlockingQueue.java:549)
        at java.util.concurrent.ExecutorCompletionService.take(ExecutorCompletionService.java:193)
        at org.terasology.world.chunks.pipeline.ChunkProcessingPipeline.chunkTaskHandler(ChunkProcessingPipeline.java:103)
        at java.lang.Thread.run(Thread.java:745)
21:30:00.529 [Chunk-Processing-3] ERROR o.t.p.i.AbstractStorageManager - Failed to load chunk zip saves\Game 14\worlds\main\-1.-1.0.chunks.zip
java.nio.channels.ClosedByInterruptException: null
        at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202)
        at sun.nio.ch.FileChannelImpl.size(FileChannelImpl.java:315)
        at com.sun.nio.zipfs.ZipFileSystem.findEND(ZipFileSystem.java:963)
        at com.sun.nio.zipfs.ZipFileSystem.initCEN(ZipFileSystem.java:1030)
        at com.sun.nio.zipfs.ZipFileSystem.<init>(ZipFileSystem.java:130)
        at com.sun.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:139)
        at java.nio.file.FileSystems.newFileSystem(FileSystems.java:390)
        at org.terasology.persistence.internal.AbstractStorageManager.loadChunkZip(AbstractStorageManager.java:127)
        at org.terasology.persistence.internal.AbstractStorageManager.loadCompressedChunk(AbstractStorageManager.java:156)
        at org.terasology.persistence.internal.ReadWriteStorageManager.loadCompressedChunk(ReadWriteStorageManager.java:349)
        at org.terasology.persistence.internal.AbstractStorageManager.loadChunkStore(AbstractStorageManager.java:108)
        at org.terasology.world.chunks.localChunkProvider.LocalChunkProvider.processReadyChunk(LocalChunkProvider.java:204)
        at org.terasology.world.chunks.pipeline.stages.ChunkTaskProvider.lambda$null$1(ChunkTaskProvider.java:41)
        at org.terasology.world.chunks.pipeline.stages.SingleChunkTask.apply(SingleChunkTask.java:44)
        at org.terasology.world.chunks.pipeline.stages.SingleChunkTask.apply(SingleChunkTask.java:17)
        at org.terasology.world.chunks.pipeline.ChunkProcessingPipeline.lambda$runTask$4(ChunkProcessingPipeline.java:180)
        at org.terasology.world.chunks.pipeline.ChunkProcessingPipeline$PositionalCallable.call(ChunkProcessingPipeline.java:320)
        at org.terasology.world.chunks.pipeline.ChunkProcessingPipeline$PositionalCallable.call(ChunkProcessingPipeline.java:305)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.terasology.world.chunks.pipeline.PositionFuture.run(PositionFuture.java:29)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
21:30:00.691 [main] ERROR o.t.l.debug.ChunkEventErrorLogger - Unload event for not loaded chunk (-12, 0, 12)
21:30:00.729 [main] ERROR o.t.l.debug.ChunkEventErrorLogger - Unload event for not loaded chunk (-11, -1, 12)
21:30:00.787 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
21:30:00.802 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
21:30:00.813 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Player65978
21:30:00.820 [main] INFO  o.t.logic.console.ConsoleImpl - [NOTIFICATION] Player "Player65978" has left the game
21:30:00.820 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
21:30:00.820 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelWorldSystem.onNewChunk(VoxelWorldSystem.java:151)
        at org.terasology.physics.engine.VoxelWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:30:00.821 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelLiquidWorldSystem.onNewChunk(VoxelLiquidWorldSystem.java:152)
        at org.terasology.physics.engine.VoxelLiquidWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:30:00.821 [main] WARN  o.t.r.world.RenderableWorldImpl - Warning: onChunkLoaded called for a null chunk!
21:30:00.821 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelWorldSystem.onNewChunk(VoxelWorldSystem.java:151)
        at org.terasology.physics.engine.VoxelWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:30:00.821 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
        at org.terasology.physics.engine.VoxelLiquidWorldSystem.onNewChunk(VoxelLiquidWorldSystem.java:152)
        at org.terasology.physics.engine.VoxelLiquidWorldSystemMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.process(EventSystemImpl.java:111)
        at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:126)
        at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:494)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:415)
        at org.terasology.engine.Terasology.main(Terasology.java:174)
21:30:00.821 [main] WARN  o.t.r.world.RenderableWorldImpl - Warning: onChunkLoaded called for a null chunk!
21:30:00.940 [Engine-Task-Pool] INFO  o.t.rendering.opengl.ScreenGrabber - Screenshot saved to saves\Game 14\previews\1.jpg!
21:30:01.617 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
21:30:01.645 [Thread-4] INFO  o.t.e.s.d.DiscordRPCSubSystem - Discord RPC >> Disconnected!
```

**Comment by Cervator** [comment]
Third attempt to test MR went perfectly fine, simply walking around normally without `ghost` or `hspeed` so it is probably something about forcing chunk/world generation to rush along too much.

**Comment by Cervator** [comment]
Closing as fixed by #4368 - cannot reproduce after merging that, hooray!
